### PR TITLE
[precompile] Implement the dynamo tracer for torch.compiler.precompile

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #195284
* #195215
* #195070
* #195014
* #194966
* #194952
* #194951
* #194950
* #194949
* #194948
* #194677
* #194622
* #194619
* #194549
* #194542
* #194541
* #194534
* #194488
* #194479
* #194455
* #194454
* #194452
* #194450
* #194449
* #194428
* #194427
* #194426
* #194409
* #194394
* #194393
* #194390
* #194389
* #194388
* #194387
* #194386
* #194385
* #194384
* #194383
* #194320
* #194319
* #194318
* #194317
* #194316
* #194302
* #194301
* #194300
* #194286
* #194285
* #194284
* #194283
* #194282
* #194250
* #194249
* #194248
* #194158
* __->__ #192379
* #193976
* #193697
* #192374
* #192866

Ahead-of-time capture of a compiled computation into an artifact you can ship:
a self-contained Python source string plus a companion cache, the
``(python_code, cache)`` pair, reloaded with
``torch.compiler.precompile.load``. One contract for every capture shape.

    code, cache = torch.compiler.precompile(fn, example_inputs=[(model, x)])
    code, cache = torch.compiler.precompile(          # graph breaks, recompiles
        fn, example_inputs=[(model, x1), (model, x2)], tracer="dynamo")
    served = torch.compiler.precompile.load(code, cache)
    served(model, x)

``torch.compiler.precompile`` and its ``(python_code, cache)`` contract already
exist; ``tracer="dynamo"`` is the value the API declares as planned and rejects
with NotImplementedError. This implements it.

It also makes ``example_inputs`` the ONLY calling convention, which is the other
half of what makes the tracer legible. precompile on main takes the example call
positionally -- and names that varargs parameter ``example_inputs`` -- so adding
a real ``example_inputs`` list beside it left two things with one name and a
``tracer`` argument that was load-bearing in one spelling and rejected in the
other. Now there is one convention and ``tracer`` is the only thing that selects
a front-end: ``"make_fx"`` (the default) is one non-strict ATen trace and takes
exactly one call, ``"dynamo"`` takes as many as you give it. A positional call
raises a TypeError naming the replacement. This is a BC break to an API that is
already on main; precompile is new enough that the blast radius is the tests.

The dynamo tracer keeps real guards and records every frame Dynamo compiled --
the entry frame plus each graph-break continuation, and every recompiled variant
of each -- into one artifact format. That is the shape aot_compile cannot express
at all: it produces exactly one graph and rejects graph breaks, so a model that
breaks 50 times and specializes 3 ways per frame cannot be precompiled today.

Collapsing the two spellings also settles the grad mode, which they used to
answer differently: the caller wrote the positional call so it ran in the
ambient mode, while precompile makes the ``example_inputs`` calls so it picked
``no_grad``. precompile now always makes the call, so ``training`` decides for
both tracers -- ``no_grad`` unless ``training=True``. For make_fx that is a
no-op on inference (the artifact is byte-identical either way, measured) and a
fix otherwise: ``training`` was previously accepted and silently ignored there.

The dynamo tracer mirrors ``torch.compile``. Where they used to differ it was
because a single-graph emitter worked around torch.compile locally -- an ambient
torch_function mode, for instance, which torch.compile applies twice and that
emitter applied once. Parity is the contract now, so those divergences are gone
and the make_fx path keeps the stronger guarantees.

Only the guards that DISCRIMINATE are serialized, and that is most of what makes
a real model viable rather than an optimization on top. Most guards cannot be
serialized at all -- identity guards on modules, callables, closure cells -- and
of those that can, the overwhelming majority are invariant: on a 62-frame
ranking model, 738 of its 1207 guard slots held the same fact in every captured
variant. Such a guard selects nothing. The contract precompile already asks for
is what licenses dropping it: the environment is identical at capture and at
runtime, and every variation in computation comes from the inputs, so an
invariant guard is either environment, pinned by that assumption, or an input
dimension the examples deliberately did not vary. The cost is real and is stated
in the code: a call outside the captured domain is served by a captured graph
rather than refused. Guards Dynamo never models -- SHAPE_ENV, GLOBAL_STATE and
the rest of _UNMODELLED_GUARD_TYPES -- are exempt, since they were never shown
to be constant, only never analyzed.

A slot is (guard_type, normalized source), and it is kept when two variants of
its frame recorded different facts for it OR when it is present in some variants
and absent in others. That second case is not a refinement, it is most of the
answer: 225 of the 274 slots kept, because a guard only one variant carries is
exactly what tells that variant apart. Which guards discriminate is knowable
only once every variant exists, and guards are serialized per compilation as
each is produced, so capture runs the examples twice: once to learn, then again
keeping only what varied. The passes agree by construction rather than by luck,
since PrecompileSession gives each session a fresh PGO state and enters under
_use_code_state, so the second cannot inherit the first's automatic-dynamic
learning.

How the multi-graph artifact serves depends on what it can reach, and the
artifact says which in its header. A source driver dispatches the entry frame
and the continuations the entry's bytecode names. When that covers every
captured frame -- the whole computation is one function that breaks a few times
-- the artifact is ``standalone``: it installs nothing, no frame evaluator is
involved, and the driver rebuilds each frame's guard trees and dispatches by
evaluating guards against the call.

A model that breaks inside nested module forwards is the other case, and it is
the common one. Those frames are entered by an ordinary call, so no name in the
entry reaches them and a source driver would run them eager -- correct, but with
every compiled variant unused. Such a capture is ``installed`` instead: the
driver rebuilds the package and hands the frames to the frame evaluator, which
reaches them by code object. ``load`` still installs nothing; entering the
result, or calling it, installs, and ``unload``/exit takes it back out, so the
mutation is attributable to a point in the caller's control flow.

Either way a call no captured variant covers RAISES rather than silently
compiling a new one, and reachability is computed as a closure: a continuation
counts as covered only once something that is itself reachable names it.

The emitted file is labelled, and the split is by what actually has a source
form. What was captured -- the frames, their variant counts, the guards that
could not be serialized, the values pinned to what capture saw -- is emitted as
readable literals. The compiled subgraphs are RENDERED AS SOURCE, kernels and
all, through the same aot_autograd.compile_to_python the make_fx tracer uses:
they are Inductor output, not Dynamo state, so pickling them was a default
rather than a constraint, and it was most of the file. On a graph-breaking model
with two shape variants that is 379,657 chars -> 102,950 with 8 readable
@triton.jit kernels; on a two-variant toy, 197,843 -> 54,632. Only the guard
trees and the transformed bytecode, which genuinely have no source form, remain
under an OPAQUE banner.

Rendering is per subgraph and the blocks share one namespace, which is the whole
difficulty. Two variants of one frame are the same computation at different
shapes, so they render the SAME top-level names, and a block resolves its
siblings as late-bound globals -- its ``call`` looks up ``_runtime_wrapper`` and
``_inner_call``, its ``Runner`` looks up the kernels. Left alone, the batch-2
variant silently runs the batch-4 variant's code. Every top-level name a block
defines is therefore suffixed per subgraph, driven by AST positions rather than
a text match: that is what keeps an attribute (``runner.call``), a nested
binding (``def call`` inside ``class Runner``) and a dotted import path
(``torch._inductor.async_compile``) out of the rewrite, each of which broke a
regex first.

A training capture keeps the bundle instead, and that is refused up front rather
than caught: compile_to_python pins AOTAutograd to the inference path, so a
training graph RENDERS SUCCESSFULLY with the backward silently gone and the
served output missing its grad_fn. The eager backend keeps it too (its backends
are fx graphs with nothing to render), as does any single subgraph whose
lowering fails; the header reports how many fell back.

The public surface is deliberately two calls. ``precompile()`` runs the examples
itself and returns the pair; ``precompile.load()`` reads it back. There is no
capture object the caller drives by hand: with ``training=True`` covering the
case that used to need one, a session would only be a second way to say the same
thing, and it would be the way that requires a loss.

Review it in five pieces: the programming-model contract in Note [precompile
programming model]; the make_fx emitter and its drivers; the
multi-graph capture session, its two passes and its gates; the standalone
dispatcher; and the public surface in torch/compiler.

``training=True`` makes it a drop-in for a training step. The examples then run
with grad enabled and the backward is lowered eagerly, so the artifact carries
one: a served output arrives with grad_fn=CompiledFunctionBackward and
``.backward()`` runs precompiled code. No loss is required for that, and none is
asked for -- AOTAutograd's joint trace already synthesizes tangents from the
forward outputs' own metadata, and already wraps the result in a
torch.autograd.Function. All that was missing is that it lowers the backward
lazily, on the first ``.backward()`` call, so a capture that never makes one
records a forward and nothing else. Two serialization bugs sat on that path and
are fixed here, both reachable only once there are grads to serialize: a
parameter's .grad is read by no guard, so it is pruned to the _Missing sentinel
and _unpickle_tensor then assigned the sentinel and raised at LOAD, and the
reducer read .grad off non-leaf tensors, which is always None and warns.

One fix in guards.py is required rather than incidental. Value pruning keys off
the guard tree: anything the tree does not reach is replaced by a placeholder.
Dropping a guard therefore dropped the VALUE it named while the rest of the
state still referred to it, and the artifact failed to load with
"empty_strided(): argument 'dtype' must be torch.dtype, not _Missing". Pruning
now follows the unfiltered tree. Note this only masks the underlying issue:
missing_values is keyed by id, so one unguarded attribute holding torch.bfloat16
can stand in for every reference to that dtype in the pickle. That is a latent
bug of its own and wants a separate fix.

Three things worth knowing before relying on an artifact. Capture is by
execution, so an unexercised path is simply absent and a complete summary means
complete for the calls that ran. A bare nn.Module is refused: Dynamo compiles
its own wrapper frame, which closes over the module, so the entry frame holds no
graph and no artifact can rebuild that closure -- capture the function that
calls the model instead, which the error says. And an installed artifact
rebuilds its entry from that entry's code object, which cannot restore a
closure, so a closure entry is likewise refused at capture, where the closure is
visible, rather than on the serving machine.

Test Plan:

```
python test/test_precompile.py
python test/dynamo/test_package.py
python test/dynamo/test_guard_serialization.py
python test/dynamo/test_recompile_ux.py
python test/dynamo/test_aot_compile.py
python test/dynamo/test_misc.py
```

297 precompile tests. The API rework rewrote 171 call sites mechanically; 34
tests then failed with "no captured variant matches", which is the grad-mode
change above surfacing -- those captures had been relying on the ambient mode --
and they now say ``training=True`` explicitly.

Automatic dynamic is covered across graph breaks: a dim that varies is detected
per frame, so a continuation whose input varied is promoted to dynamic while an
entry frame whose input never did stays specialized, and unseen shapes then
serve. The graph-break suite runs each of six break shapes -- a disabled callee,
a disabled method, explicit breaks, a data-dependent break, a break in a child
module, a break inside a loop -- on both backends and asserts the served result
equals what torch.compile returns for the same call.

On the 62-frame ranking model that motivated this: 738 of 1207 guard slots
dropped, all five captured calls exact with zero graphs compiled while serving,
capture 5.6s for one pass and 10.3s for the two.

Authored with an AI assistant.

Differential Revision: [D116222523](https://our.internmc.facebook.com/intern/diff/D116222523)

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222523).

**Comment by bobrenjc93** [comment]
Superseded by the canonical precompile stack #195353-#195362; this change lives in #195360, #195361 and #195362. Closing to consolidate review there. The branch is kept on GitHub.
